warehouse/oidc/models/activestate.py

```diff
@@ -0,0 +1,204 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import urllib
+
+from typing import Any
+
+from sqlalchemy import ForeignKey, String, UniqueConstraint
+from sqlalchemy.dialects.postgresql import UUID
+from sqlalchemy.orm import Query, mapped_column
+
+import warehouse.oidc.models._core as oidccore
+
+from warehouse.oidc.errors import InvalidPublisherError
+from warehouse.oidc.interfaces import SignedClaims
+from warehouse.oidc.models._core import (
+    CheckClaimCallable,
+    OIDCPublisher,
+    PendingOIDCPublisher,
+)
+
+ACTIVESTATE_URL = "https://platform.activestate.com"
+
+ACTIVESTATE_OIDC_API_PATH = "api/v1/oauth/oidc"
+
+
+def _check_sub(
+    ground_truth: str, signed_claim: str, _all_signed_claims: SignedClaims
+) -> bool:
+    # We expect a string formatted as follows:
+    #  org:<orgName>:project:<projectName>
+    # where :OPTIONAL-STUFF is a concatenation of other job context
+    # metadata. We currently lack the ground context to verify that
+    # additional metadata, so we limit our verification to just the ORG/REPO
+    # component.
+
+    # Defensive: ActiveState should never give us an empty subject.
+    if not signed_claim:
+        raise InvalidPublisherError("Missing 'subject' claim")
+
+    components = signed_claim.split(":")
+
+    if len(components) < 4:
+        raise InvalidPublisherError("Invalid 'subject' claim. Wrong format")
+
+    matches = (
+        f"{components[0]}:{components[1]}:{components[2]}:{components[3]}"
+        == ground_truth
+    )
+    if not matches:
+        raise InvalidPublisherError("Invalid 'subject' claim")
+
+    return True
+
+
+class ActiveStatePublisherMixin:
+    """
+    Common functionality for both pending and concrete ActiveState OIDC publishers.
+    """
+
+    organization = mapped_column(String, nullable=False)
+    activestate_project_name = mapped_column(String, nullable=False)
+    actor = mapped_column(String, nullable=False)
+    # 'actor' (The ActiveState platform username) is obstained from the user
+    # while configuring the publisher We'll make an api call to ActiveState to
+    # get the 'actor_id'
+    actor_id = mapped_column(String, nullable=False)
```

Comment:
Do we need to do the same thing for `organization` that we do for `actor`/`actor_id` here? (i.e., get the `organization_id` and compare against that instead).

Reply:
I wasn't intending to, no.  In the form PR, I confirm the organization exists with an API call but I don't save anything: https://github.com/pypi/warehouse/pull/15168/files#diff-7d755d844461cdc27241084bdffddb5860d65107d1ac79d7a0e56c6b4c7d8f11R122

Reply:
I'm not familiar enough with Activestate: if a user deletes their organization named `foo`, can another user resurrect the same organization? I guess if we're also validating that the `actor_id` has not changed, even if that was possible this should be OK, but just want to validate my assumptions.

Reply:
Ahh I understand and great point.  We don't allow re-use of deleted organization names due to possible name squatting.   And yeah, as you said, we're validating the user with the `actor_id`.

Reply:
Great, thanks for clarifying! Resolving this.